src/wheredataframe/wheredataframe.jl

```diff
@@ -0,0 +1,287 @@
+"""
+    WhereDataFrame{<:AbstractDataFrame,<:AbstractIndex}
+
+    The result of a [`where`](@ref) operation on an `AbstractDataFrame`; a
+    subset of a `AbstractDataFrame`
+    Not meant to be constructed directly, see `where`.
+"""
+struct WhereDataFrame{D<:AbstractDataFrame, T<:AbstractVector{Int}}
+    parent::D
+    rows::T
+end
+
+function WhereDataFrame(parent::AbstractDataFrame, rows::AbstractVector{Bool})
+    if length(rows) != nrow(parent)
+        throw(ArgumentError("invalid length of `AbstractVector{Bool}` row index" *
+                            " (got $(length(rows)), expected $(nrow(parent)))"))
+    end
+    return WhereDataFrame(parent, findall(rows))
+end
+
+rows(wdf::WhereDataFrame) = getfield(wdf, :rows)
+Base.parent(wdf::WhereDataFrame) = getfield(wdf, :parent)
+
+
+"""
+    where(d::AbstractDataFrame, args...)
+
+`where` introduces a `where` clause, which will be applied in the next function. Returns a `WhereDataFrame`.
+
+`args...` obey the same syntax as `select(d, args...)`
+Rows that return missing are understood as false
+
+- `filter`/`filter!` returns an AbstractDataFrame after filtering (resp. deleting) specified rows
+- `transform/transform!` returns an AbstractDataFrame with as many rows as the original `AbstractDataFrame` after applying the transformation on specified rows
+- `combine` and `describe` return the same thing as the function applied to a view of the `AbstractDataFrame`
+
+# Examples
+```julia
+julia> df = DataFrame(a = repeat([1, 2, 3, missing], outer=[2]),
+               b = repeat([2, 1], outer=[4]),
+               c = randn(8))
+
+# filter rows that satisfies a certain condition
+julia> filter(where(df, :a => ByRow(>(1))))
+julia> filter(where(df, :a => ByRow(>(1)), :b => x -> x .< 2))
+
+# transform only certain rows
+julia> transform(where(df, :a => ByRow(ismissing)), :a => (x -> 0) => :a)
+julia> transform(where(df, :a => ByRow(!ismissing)), :a => cumsum)
+# combine using certain rows
+julia> combine(where(df, :a => ByRow(!ismissing)), :a => sum)
+```
+"""
+function where(df::AbstractDataFrame, args...)
+    dfr = select(df, args...)
+    if any(x -> !(eltype(x) <: Union{Bool, Missing}), eachcol(dfr))
+        throw("Conditions do not evaluate to bool or missing")
+    end
+    if size(dfr, 2) == 1
+        WhereDataFrame(df, coalesce.(dfr[!, 1], false))
+    else
+        WhereDataFrame(df, coalesce.(.&(eachcol(dfr)...), false))
+    end
+end
+
+##############################################################################
+##
+## Show: show rows of parent that satisfies the `where` condition
+## with original row number
+##
+##############################################################################
+function Base.summary(wdf::WhereDataFrame)
+    @sprintf("%d×%d %s", size(parent(wdf))..., "WhereDataFrame")
+end
+
+Base.summary(io::IO, wdf::WhereDataFrame) = summary(wdf)
+
+
+function Base.show(io::IO,
```

Comment:
All mime types should be supported (in particular HTML and LaTeX). Also I would make sure that we very clearly visually indicate that this is NOT an `AbstractDataFrame`.

Finally - standard `show` has to be updated to avoid `StackOverflow` when trying to display a data frame with `WhereDataFrame` in a circular reference.

Reply:
Also in general - I would wait with merging this PR after PrettyTables.jl integration is done and then update it accordingly. c.f. #2429